File: public/cpp/051-climbing-stairs.cpp
/*
 * LeetCode Problem #70: Climbing Stairs
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/climbing-stairs/
 * 
 * Problem Statement:
 * You are climbing a staircase. It takes n steps to reach the top.
 * Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
 * 
 * Example:
 * Input: n = 3
 * Output: 3
 * Explanation: There are three ways to climb to the top:
 * 1. 1 step + 1 step + 1 step
 * 2. 1 step + 2 steps
 * 3. 2 steps + 1 step
 */

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Dynamic Programming (Bottom-Up) ====================
    /*
     * Algorithm:
     * - Build solution from base cases up to n
     * - dp[i] = ways to reach step i
     * - dp[i] = dp[i-1] + dp[i-2]
     * - Base: dp[0] = 1, dp[1] = 1
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     * 
     * When to use: Standard DP approach
     * 
     * Key Insight:
     * - To reach step i, come from step i-1 or i-2
     * - Total ways = ways from i-1 + ways from i-2
     * - This is Fibonacci sequence!
     */
    int climbStairs(int n) {
        if (n <= 2) return n;
        
        vector<int> dp(n + 1);
        dp[0] = 1;  // 1 way to stay at ground
        dp[1] = 1;  // 1 way to reach step 1
        
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i-1] + dp[i-2];
        }
        
        return dp[n];
    }
    
    // ==================== APPROACH 2: Space-Optimized DP ====================
    /*
     * Algorithm:
     * - Only need last two values, not entire array
     * - Use two variables instead of array
     * - Update variables in each iteration
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - Only previous two states needed
     * - Constant space optimization
     * - Same logic as Fibonacci
     */
    int climbStairs_Optimized(int n) {
        if (n <= 2) return n;
        
        int prev2 = 1;  // dp[i-2]
        int prev1 = 1;  // dp[i-1]
        int current;
        
        for (int i = 2; i <= n; i++) {
            current = prev1 + prev2;
            prev2 = prev1;
            prev1 = current;
        }
        
        return current;
    }
    
    // ==================== APPROACH 3: Recursive with Memoization (Top-Down) ====================
    /*
     * Algorithm:
     * - Recursively calculate ways from top
     * - Cache results to avoid recomputation
     * - Base cases: n=0 returns 1, n=1 returns 1
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(n) for memo + recursion stack
     * 
     * When to use: When recursive thinking is natural
     * 
     * Key Insight:
     * - Top-down approach with memoization
     * - Avoids exponential time of naive recursion
     */
    int climbStairs_Memo(int n) {
        vector<int> memo(n + 1, -1);
        return climbHelper(n, memo);
    }
    
private:
    int climbHelper(int n, vector<int>& memo) {
        if (n <= 1) return 1;
        
        if (memo[n] != -1) {
            return memo[n];
        }
        
        memo[n] = climbHelper(n-1, memo) + climbHelper(n-2, memo);
        return memo[n];
    }
    
public:
    // ==================== APPROACH 4: Matrix Exponentiation ====================
    /*
     * Algorithm:
     * - Use matrix exponentiation for Fibonacci
     * - [[1,1],[1,0]]^n gives Fibonacci numbers
     * - Fast exponentiation in O(log n)
     * 
     * Time Complexity: O(log n)
     * Space Complexity: O(1)
     * 
     * When to use: For very large n (advanced)
     * 
     * Key Insight:
     * - Fibonacci can be computed via matrix power
     * - Binary exponentiation reduces time
     * - Overkill for this problem but interesting
     */
    int climbStairs_Matrix(int n) {
        if (n <= 2) return n;
        
        vector<vector<long long>> base = {{1, 1}, {1, 0}};
        vector<vector<long long>> result = matrixPower(base, n);
        
        return result[0][0];
    }
    
private:
    vector<vector<long long>> matrixMultiply(vector<vector<long long>>& A, 
                                             vector<vector<long long>>& B) {
        vector<vector<long long>> C(2, vector<long long>(2));
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                C[i][j] = 0;
                for (int k = 0; k < 2; k++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return C;
    }
    
    vector<vector<long long>> matrixPower(vector<vector<long long>>& matrix, int n) {
        if (n == 1) return matrix;
        
        if (n % 2 == 0) {
            auto half = matrixPower(matrix, n / 2);
            return matrixMultiply(half, half);
        } else {
            auto half = matrixPower(matrix, n - 1);
            return matrixMultiply(matrix, half);
        }
    }
    
public:
    // ==================== APPROACH 5: Mathematical Formula (Binet's Formula) ====================
    /*
     * Algorithm:
     * - Use closed-form Fibonacci formula
     * - F(n) = (φ^n - ψ^n) / √5
     * - φ = (1 + √5) / 2, ψ = (1 - √5) / 2
     * 
     * Time Complexity: O(1)
     * Space Complexity: O(1)
     * 
     * When to use: For mathematical elegance (not practical)
     * 
     * Note: Floating point precision issues for large n
     */
    int climbStairs_Formula(int n) {
        double sqrt5 = sqrt(5);
        double phi = (1 + sqrt5) / 2;
        double psi = (1 - sqrt5) / 2;
        
        return round((pow(phi, n + 1) - pow(psi, n + 1)) / sqrt5);
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Small n
    // Input: n = 2
    // Expected: 2 (1+1, 2)
    cout << "Test Case 1: n = 2\n";
    cout << "Output: " << sol.climbStairs(2) << "\n";
    cout << "Expected: 2\n\n";
    
    // Test Case 2: Medium n
    // Input: n = 3
    // Expected: 3 (1+1+1, 1+2, 2+1)
    cout << "Test Case 2: n = 3\n";
    cout << "Output: " << sol.climbStairs_Optimized(3) << "\n";
    cout << "Expected: 3\n\n";
    
    // Test Case 3: Larger n
    // Input: n = 5
    // Expected: 8
    cout << "Test Case 3: n = 5\n";
    cout << "Output: " << sol.climbStairs_Memo(5) << "\n";
    cout << "Expected: 8\n\n";
    
    // Test Case 4: Base case
    // Input: n = 1
    // Expected: 1
    cout << "Test Case 4: n = 1\n";
    cout << "Output: " << sol.climbStairs_Matrix(1) << "\n";
    cout << "Expected: 1\n\n";
    
    // Test Case 5: Larger value
    // Input: n = 10
    // Expected: 89
    cout << "Test Case 5: n = 10\n";
    cout << "Output: " << sol.climbStairs_Formula(10) << "\n";
    cout << "Expected: 89\n\n";
    
    // Test Case 6: Verify Fibonacci pattern
    cout << "Test Case 6: Fibonacci pattern verification\n";
    cout << "n=1: " << sol.climbStairs(1) << " (Expected: 1)\n";
    cout << "n=2: " << sol.climbStairs(2) << " (Expected: 2)\n";
    cout << "n=3: " << sol.climbStairs(3) << " (Expected: 3)\n";
    cout << "n=4: " << sol.climbStairs(4) << " (Expected: 5)\n";
    cout << "n=5: " << sol.climbStairs(5) << " (Expected: 8)\n";
    cout << "n=6: " << sol.climbStairs(6) << " (Expected: 13)\n\n";
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. DP Bottom-Up:
 *    Time: O(n), Space: O(n)
 *    Pros: Clear, easy to understand
 *    Cons: Uses O(n) space
 *    Best for: Learning DP concepts
 * 
 * 2. Space-Optimized DP (RECOMMENDED):
 *    Time: O(n), Space: O(1)
 *    Pros: Optimal space, simple
 *    Cons: None
 *    Best for: Most interview scenarios
 * 
 * 3. Memoization Top-Down:
 *    Time: O(n), Space: O(n)
 *    Pros: Intuitive recursion
 *    Cons: Extra space, recursion overhead
 *    Best for: When recursion is natural
 * 
 * 4. Matrix Exponentiation:
 *    Time: O(log n), Space: O(1)
 *    Pros: Fastest for very large n
 *    Cons: Complex, overkill for this problem
 *    Best for: Advanced optimization
 * 
 * 5. Mathematical Formula:
 *    Time: O(1), Space: O(1)
 *    Pros: Constant time
 *    Cons: Floating point precision issues
 *    Best for: Mathematical elegance only
 * 
 * INTERVIEW TIPS:
 * - Recognize this as Fibonacci sequence
 * - Start with recursive solution, then optimize
 * - Explain why dp[i] = dp[i-1] + dp[i-2]
 * - Mention space optimization opportunity
 * - Draw small examples (n=1,2,3) to verify
 * - Discuss base cases carefully
 * - Consider edge cases: n=0, n=1
 * - Mention this is classic DP problem
 * - Discuss how to extend to k steps
 * - Consider follow-up: count ways with specific constraints
 * 
 * KEY INSIGHTS:
 * - This is Fibonacci sequence: F(n+1)
 * - To reach step n: come from n-1 or n-2
 * - Overlapping subproblems → DP applicable
 * - Only need last two values → space optimization
 * - Pattern: 1, 2, 3, 5, 8, 13, 21...
 * - Can be generalized to k steps per move
 * 
 * STEP-BY-STEP WALKTHROUGH:
 * n = 5
 * 
 * dp[0] = 1 (base)
 * dp[1] = 1 (base)
 * dp[2] = dp[1] + dp[0] = 1 + 1 = 2
 * dp[3] = dp[2] + dp[1] = 2 + 1 = 3
 * dp[4] = dp[3] + dp[2] = 3 + 2 = 5
 * dp[5] = dp[4] + dp[3] = 5 + 3 = 8
 * 
 * Ways to reach step 5: 8
 * 
 * Actual paths:
 * 1. 1+1+1+1+1
 * 2. 1+1+1+2
 * 3. 1+1+2+1
 * 4. 1+2+1+1
 * 5. 2+1+1+1
 * 6. 1+2+2
 * 7. 2+1+2
 * 8. 2+2+1
 * 
 * COMMON MISTAKES:
 * - Starting with dp[0] = 0 (should be 1)
 * - Confusing n with array index
 * - Not handling n=1 separately
 * - Forgetting base cases
 * - Off-by-one errors in loop
 * - Not recognizing Fibonacci pattern
 * - Using naive recursion (exponential time)
 * - Integer overflow for large n
 * - Incorrect space optimization
 * - Not considering n=0 case
 * 
 * FOLLOW-UP QUESTIONS:
 * - What if you can climb 1, 2, or 3 steps? (generalize to k steps)
 * - What if some steps are broken and cannot be used?
 * - What if there's a cost associated with each step?
 * - How would you find the minimum cost path?
 * - How would you print all possible paths?
 * 
 * RELATED PROBLEMS:
 * - LeetCode #746: Min Cost Climbing Stairs
 * - LeetCode #509: Fibonacci Number
 * - LeetCode #1137: N-th Tribonacci Number
 * - LeetCode #91: Decode Ways (similar DP pattern)
 * - LeetCode #198: House Robber (similar DP)
 */

// Made with Bob

File: src/data/tutorials/051-climbing-stairs.js
/**
 * Tutorial content for #051 — Climbing Stairs
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `You are climbing a staircase with n steps. Each time you can either climb 1 step or 2 steps. In how many distinct ways can you reach the top?`,
    example: `n = 4\n→ (1+1+1+1), (1+1+2), (1+2+1), (2+1+1), (2+2)\n✅ Answer: 5`,
    keyInsight: `To reach step n you must have come from step n-1 (took 1 step) or step n-2 (took 2 steps). So ways(n) = ways(n-1) + ways(n-2) — the Fibonacci recurrence.`,
  },

  approaches: {
    'Recursive Memoization (Top-Down DP)': {
      intuition: `Think recursively: to count ways to reach step n, count ways to reach step n-1 and add it to ways to reach step n-2. Without caching, the same sub-problems are solved over and over. Memoisation stores each result the first time it is computed so every step is solved exactly once.`,
      steps: [
        `Create a memo array of size n+1 filled with -1 (unsolved).`,
        `Base cases: ways(0) = 1, ways(1) = 1.`,
        `For any other n: check memo first; if filled, return it.`,
        `Otherwise compute memo[n] = solve(n-1) + solve(n-2).`,
        `Return memo[n].`,
      ],
      example: `n = 4\n\nsolve(4) = solve(3) + solve(2)\nsolve(3) = solve(2) + solve(1)\nsolve(2) = solve(1) + solve(0) = 1 + 1 = 2  ← cached\nsolve(3) = 2 + 1 = 3  ← cached\nsolve(4) = 3 + 2 = 5\n✅ Answer: 5`,
      keyInsight: `O(n) time and O(n) space. The memo array turns an exponential recursion tree into a linear chain of lookups.`,
    },

    'Bottom-Up DP': {
      intuition: `Build the answer from the smallest steps upward. dp[i] stores the number of ways to reach step i. We fill it left-to-right using dp[i] = dp[i-1] + dp[i-2]. No recursion or stack needed.`,
      steps: [
        `Create a dp array of size n+1.`,
        `Set dp[0] = 1 and dp[1] = 1.`,
        `For i from 2 to n: dp[i] = dp[i-1] + dp[i-2].`,
        `Return dp[n].`,
      ],
      example: `n = 4\n\ndp[0] = 1\ndp[1] = 1\ndp[2] = dp[1] + dp[0] = 2\ndp[3] = dp[2] + dp[1] = 3\ndp[4] = dp[3] + dp[2] = 5\n✅ Answer: 5`,
      keyInsight: `O(n) time and O(n) space. Iterative so no risk of stack overflow on large n.`,
    },

    'Space-Optimized DP': {
      intuition: `The bottom-up recurrence only looks back two positions. Instead of storing the whole array, keep just two variables — one for the previous step and one for the step before that — and slide them forward.`,
      steps: [
        `Set prev2 = 1 (ways to reach step 0) and prev1 = 1 (ways to reach step 1).`,
        `For each step i from 2 to n:`,
        `  cur = prev1 + prev2`,
        `  prev2 = prev1`,
        `  prev1 = cur`,
        `Return prev1.`,
      ],
      example: `n = 4\n\nprev2=1, prev1=1\ni=2: cur=1+1=2 → prev2=1, prev1=2\ni=3: cur=2+1=3 → prev2=2, prev1=3\ni=4: cur=3+2=5 → prev2=3, prev1=5\nReturn 5\n✅ Answer: 5`,
      keyInsight: `O(n) time and O(1) space — the interview-optimal solution. Only two variables are ever needed because the recurrence has a window of two.`,
    },

    'Matrix Exponentiation': {
      intuition: `The Fibonacci recurrence can be written as a matrix multiplication: [f(n), f(n-1)] = [[1,1],[1,0]]^n × [1, 0]. We can compute the matrix raised to the power n using fast exponentiation (repeated squaring), halving the number of multiplications.`,
      steps: [
        `Represent the transition as the 2×2 matrix M = [[1,1],[1,0]].`,
        `Use fast matrix exponentiation: M^n via repeated squaring.`,
        `Start with result = identity matrix, base = M.`,
        `While n > 0: if n is odd, result = result × base; base = base × base; n = n / 2.`,
        `Return result[0][1] (which equals fib(n)).`,
      ],
      example: `n = 4\n\nM = [[1,1],[1,0]]\nM^2 = [[2,1],[1,1]]\nM^4 = [[5,3],[3,2]]\nResult[0][1] = 3... but ways(4) = dp[4] = 5\nActually result[0][0] = 5 ✅\n✅ Answer: 5`,
      keyInsight: `O(log n) time and O(1) space. Overkill for this problem but essential when n can be astronomically large.`,
    },

    "Mathematical Formula (Binet's Formula)": {
      intuition: `The climbing-stairs sequence is exactly Fibonacci shifted by one position. Binet's formula gives Fibonacci numbers directly using the golden ratio φ = (1+√5)/2. We can compute ways(n) = round( φ^(n+1) / √5 ) in a single arithmetic expression — no loops needed.`,
      steps: [
        `Compute φ = (1 + √5) / 2 (the golden ratio, ≈ 1.618).`,
        `Compute ψ = (1 - √5) / 2 (the conjugate, ≈ -0.618).`,
        `Result = round( (φ^(n+1) - ψ^(n+1)) / √5 ).`,
        `The ψ term shrinks so fast it only matters for rounding — many implementations just use round(φ^(n+1) / √5).`,
      ],
      example: `n = 4\n\nφ ≈ 1.618\nφ^5 ≈ 11.09\n11.09 / 2.236 ≈ 4.959\nround(4.959) = 5\n✅ Answer: 5`,
      keyInsight: `O(1) time conceptually, but floating-point precision limits correctness to small n (roughly n ≤ 70). Impractical for competitive programming but a great interview talking point.`,
    },
  },
}
